// @rhoova/node-client/src/app/RhoovaClient.js
const RhoovaConfig = require('./RhoovaConfig');
const CalculationType = require('./CalculationType');
const DataType = require('./DataType');
const axios = require('axios');

class RhoovaClient {

    constructor({apiKey, apiSecret, apiUrl}) {
        this.config = new RhoovaConfig(
            apiKey,
            apiSecret,
            apiUrl || 'https://app.rhoova.com/api'
        );
    }

    createTask({calculationType, data, waitResult}) {
        if (!Object.values(CalculationType).includes(calculationType)) {
            throw new Error("'Calculation type' must be a CalculationType")
        } else {
            return new Promise((resolve, reject) => {
                axios.post(this.config.getApiUrl() + "/tasks/" + calculationType, data, {
                        headers: {
                            'Content-type': 'application/json',
                            'ekx-api-key': this.config.getApiKey(),
                            'ekx-api-secret': this.config.getApiSecret(),
                            'ekx-wait-result': waitResult ? waitResult : false
                        }
                    }
                ).then((response) => {
                    resolve(response.data);
                    // if (response.data.error) {
                    //     reject(response.data.error)
                    // } else if (response.data.result) {
                    //     resolve(JSON.stringify({
                    //         RESULT: JSON.parse(response.data.result),
                    //         TASKID: response.data.pid
                    //     }, null, 6))
                    // } else if (response.data.taskId) {
                    //     resolve({taskId: response.data.taskId})
                    // } else {
                    //     reject("Result is still calculated")
                    // }
                }).catch((error) => {
                    if (error.response) {
                        if (error.response.status >= 400 && error.response.status < 600) {
                            let err = error.response.data.errors ? error.response.data.errors[0].detail :
                                error.response.data;
                            reject({
                                error_name: error.response.data.title,
                                error_status: error.response.status,
                                error_detail: err
                            })
                        }
                    } else if (error.isAxiosError) {
                        reject("AXIOS ERROR !!")
                    } else {
                        reject(error)
                    }
                });
            });

        }
    }

    getTaskResult({taskId}) {
        if (taskId && typeof taskId !== 'string') {
            throw new Error("'taskId' must be a string")
        } else {
            let _this = this;
            return new Promise(function (resolve, reject) {
                axios.get(_this.config.getApiUrl() + "/tasks/" + taskId, {
                    headers: {
                        'Content-type': 'application/json',
                        'ekx-api-key': _this.config.getApiKey(),
                        'ekx-api-secret': _this.config.getApiSecret()
                    }
                }).then(async (response) => {
                    resolve(response.data);
                    // if (response.data.error) {
                    //     reject(response.data.error)
                    // } else if (response.data.result) {
                    //     resolve(JSON.stringify({
                    //         RESULT: JSON.parse(response.data.result),
                    //         TASKID: response.data.pid
                    //     }, null, 6))
                    // } else if (response.data.errors) {
                    //     reject(response.data.errors)
                    // } else {
                    //     reject("Results can not be calculated")
                    // }
                }).catch((error) => {
                    if (error.response) {
                        if (error.response.status >= 400 && error.response.status < 600) {
                            let err = error.response.data.errors ? error.response.data.errors[0].detail :
                                error.response.data;
                            reject({
                                error_name: error.response.data.title,
                                error_status: error.response.status,
                                error_detail: err
                            })
                        }
                    } else if (error.isAxiosError) {
                        reject("AXIOS ERROR !!")
                    } else {
                        reject(error)
                    }
                })
            });
        }

    }

    loadData({file, name, type}) {
        if (!Object.values(DataType).includes(type)) {
            throw new Error("'Type' must be a DataType")
        }
        if (file && typeof file !== 'object') {
            throw new Error("'Data type' must be an object")
        } else {
            return new Promise((resolve, reject) => {
                axios.post(this.config.getApiUrl() + "/data-sources", {
                        file: file,
                        name: name,
                        type: type
                    }, {
                        headers: {
                            'Content-type': 'application/json',
                            'ekx-api-key': this.config.getApiKey(),
                            'ekx-api-secret': this.config.getApiSecret()
                        }
                    }
                ).then((response) => {
                    resolve(JSON.stringify({
                        DATA: JSON.parse(JSON.parse(response.data.data).data),
                        NAME: response.data.name,
                        ID: response.data.pid,
                        TYPE: Object.keys(DataType)[response.data.type]
                    }, null, 6))
                }).catch((error) => {
                    if (error.response) {
                        if (error.response.status >= 400 && error.response.status < 600) {
                            let err = error.response.data.errors ? error.response.data.errors[0].detail :
                                error.response.data;
                            reject({
                                error_name: error.response.data.title,
                                error_status: error.response.status,
                                error_detail: err
                            })
                        }
                    } else if (error.isAxiosError) {
                        reject("AXIOS ERROR !!")
                    } else {
                        reject(error)
                    }
                });
            });
        }
    }

    getData({id}) {
        return new Promise((resolve, reject) => {
            axios.get(this.config.getApiUrl() + "/data-sources/" + id, {
                    headers: {
                        'Content-type': 'application/json',
                        'ekx-api-key': this.config.getApiKey(),
                        'ekx-api-secret': this.config.getApiSecret()
                    }
                }
            ).then((response) => {
                resolve(JSON.stringify({
                    DATA: JSON.parse(JSON.parse(response.data.data).data),
                    NAME: response.data.name,
                    ID: response.data.pid,
                    TYPE: Object.keys(DataType)[response.data.type]
                }, null, 6))
            }).catch((error) => {
                if (error.response) {
                    if (error.response.status >= 400 && error.response.status < 600) {
                        let err = error.response.data.errors ? error.response.data.errors[0].detail :
                            error.response.data;
                        reject({
                            error_name: error.response.data.title,
                            error_status: error.response.status,
                            error_detail: err
                        })
                    }
                } else if (error.isAxiosError) {
                    reject("AXIOS ERROR !!")
                } else {
                    reject(error)
                }
            });
        });
    }
}

module.exports = RhoovaClient;
